src/spawn_stubs.c

```diff
@@ -478,44 +506,34 @@ CAMLprim value spawn_unix(value v_env,
   sigfillset(&sigset);
   pthread_sigmask(SIG_SETMASK, &sigset, &saved_procmask);
 
-  if (v_sigprocmask == Val_long(0)) {
-    sigemptyset(&info.child_sigmask);
-  } else {
-    v_sigprocmask = Field(v_sigprocmask, 0);
-    value v_sigprocmask_command = Field(v_sigprocmask, 0);
-    enum caml_unix_sigprocmask_command sigprocmask_command = Long_val(v_sigprocmask_command);
+  switch (sigprocmask_command) {
+    case CAML_SIG_SETMASK:
+      sigemptyset(&info.child_sigmask);
+      break;
+
+    case CAML_SIG_BLOCK:
+    case CAML_SIG_UNBLOCK:
+      info.child_sigmask = saved_procmask;
+      break;
 
+    default:
+      caml_failwith("Unknown sigprocmask action");
```

Comment:
Can't call `caml_failwith` in this context -- need `caml_leave_blocking_section` (etc) first.  See the cleanup when `safe_pipe` fails above.

Reply:
Whoops, it's as if I've never read the patch. This makes me convinced that we should split the file in two files:
- the part with bindings that runs with OCaml lock held
- the part with thread-safe code that doesn't even include any of the OCaml headers

I'll prepare a feature that does that.